### PR TITLE
Adds ability to search and filter schema objects

### PR DESCRIPTION
**Pull request by garyr**
## Description
This feature allows for a user to search for and filter the list of schemas in the XDM Object view

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All tests pass and I've made any necessary test changes.
- [x] I've updated the schema in extension.json or no changes are necessary.

### CODE REVIEW

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [inline src/view/dataElements/xdmObject.jsx:168]: Rather than using `u-gapTop` here, I would recommend using `u-alignItemsCenter` on the div that contains this div. It will probably result in more accurate centering and is more robust if the size of the Wait or ComboBox changes.

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [inline src/view/dataElements/xdmObject.jsx:169]: `schemaSearchInProgress` is guaranteed to be a boolean so you can just do `schemaSearchInProgress && ...`

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [inline src/view/dataElements/xdmObject.jsx:266]: The `sandboxes.sandboxes` thing can get a little confusing (not just here but in other parts of the code). What if we did `setSandboxes(_sandboxes.sandboxes)` and then had another state called `sandboxesEnabled` that you would set using `setSandboxesEnabled(!_sandboxes.disabled)`. I'd probably also rename `_sandboxes` here to `sandboxesResult` or do `.then({ sandboxes, disabled})` [here](https://github.com/adobe/reactor-extension-alloy/pull/116/files#diff-49c61fe83e69899661207037ed001ed1f94dacac52029c9b80e9c50000c882aeR262) and then you just avoid naming the object altogether. I hope that makes sense.

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [inline src/view/dataElements/xdmObject.jsx:89]: We need to take into consideration that a new request may be sent while a previous request is still active. In other words, if there's an active request and a subsequent request is made, we need to either cancel the active request (https://davidwalsh.name/cancel-fetch) or just ignore the response to the active request.

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [inline src/view/dataElements/xdmObject.jsx:146]: In Slack you mentioned:

> There is a small issue when filtering the schema results and deleting the filter. After removing all the text search query, the combo box results do not update unless you click twice on the button to drop down the list.

I wanted to make sure you're aware that the bug might be a little more complex than this. Notice when I type "s" and then backspace and repeat the process:

searchconcurrency

Notice the results vary.

**Review by Aaronius — COMMENTED** [review]
Looking good @garyr. I've left some feedback. I agree with your assessment that xdmObject.jsx is getting bloated and complex. It will take some thought to figure out how to best break it down and simplify it. I wouldn't let it hold up your PR from being merged though.

We do need to figure out the problem with the filter when clearing out the textfield. I think @jonsnyder was thinking about looking into it, but otherwise I'll try to dig in later today.

File: src/view/dataElements/xdmObject.jsx
Comment (Aaronius) [reply src/view/dataElements/xdmObject.jsx:146]: Here's some other weird behavior. Type "s", then backspace, then "u", then backspace.

filteringproblem2

Comment (Aaronius) [reply src/view/dataElements/xdmObject.jsx:146]: I think the problem here is that the React-Spectrum ComboBox already has filtering in it (when you type in the input, ComboBox will filter down the options to just those that match), but we're also doing filtering on our side as the user types and then updating the `options` passed into ComboBox. ComboBox is getting all kinds of confused.

I'm not sure, but I think continuing with ComboBox will lead us down a long rabbit hole. I would recommend swapping it out for something like https://git.corp.adobe.com/reactor/extension-adobeanalytics/blob/master/src/view/extensionConfiguration/components/restrictedComboBox.jsx, which leverages Autocomplete but makes it look similar to the ComboBox component. Autocomplete has a `getCompletions` API, which is much more suitable for this type of use case since we won't be competing with the component on who's doing the filtering.

Comment (jonsnyder) [reply src/view/dataElements/xdmObject.jsx:146]: CompletionComboBox from the analytics extension is closer to what we need here. I am going to try and see if that fixes the issues.

File: src/view/dataElements/xdmObject.jsx
Comment (garyr) [reply src/view/dataElements/xdmObject.jsx]: I think this has been addressed with the move to the `CompletionComboBox` component. For now we have removed the search progress

File: src/view/dataElements/xdmObject.jsx
Comment (garyr) [reply src/view/dataElements/xdmObject.jsx:258]: I agree that this should probably be enhanced. I would like to address this with an upcoming change that will likely involve refactoring the sandbox feature into some shared code

File: src/view/dataElements/xdmObject.jsx
Comment (garyr) [reply src/view/dataElements/xdmObject.jsx]: I think this is addressed with the `CompletionComboBox`

File: src/view/dataElements/xdmObject.jsx
Comment (garyr) [reply src/view/dataElements/xdmObject.jsx]: No longer needed

File: src/view/dataElements/xdmObject.jsx
Comment (garyr) [reply src/view/dataElements/xdmObject.jsx]: Thanks. I made your suggested change

**Review by jonsnyder — APPROVED** [review]

**Review by Aaronius — APPROVED** [review]
* If I click the dropdown button, then click somewhere outside the combobox, it automatically selects the first item.
* The filtering is case sensitive when it should probably be case insensitive.
* Select a schema. After that, clear out the search field, then click outside of the combobox. I see this shown:

image

The looks like something went wrong, but really the user just deselected the schema they had selected. I would recommend we just down show the alert in that case.

I think case sensitivity would be great to fix before releasing, but we can probably get away with not addressing the other points.

I would recommend squashing when you hit the merge button.

**Comment by Aaronius** [comment]
Thanks for working through this Gary and Jon. I wrote my last comment in a hurry and didn't even preface it by saying "These are a few issues I noticed while poking around." Anyway, it's looking good! Thanks again!

**Comment by Aaronius** [comment]
@garyr, I looked into the TestCafe hanging problem you mentioned. When I was messing around with it, I was able to get it to hang once out of about 20 times of trying different things. I may just not have the reproduction steps down, but anyway, it seemed quite rare. From what I saw, it looked like it's probably more of a bug between TestCafe and the browser than the code we manage. I'm not sure that it's isolated to your branch either.

Anyway, I went ahead and upgraded TestCafe in case that may have any improvement. I didn't see anything that specifically related in the release notes, but there may be something there. I also merged master into this branch because Nina merge a PR that fixed a functional test and I wanted to see all tests passing on this branch.

Anyway, I think you're good to merge whenever you feel ready.

File: test/functional/helpers/spectrum.js
Comment (Aaronius) [inline test/functional/helpers/spectrum.js:144]: This was a bug that was introduced in one of the newer versions of TestCafe. It's supposed to be fixed in the next stable release.

File: test/functional/dataElements/xdmObject.spec.js
Comment (Aaronius) [inline test/functional/dataElements/xdmObject.spec.js:64]: The prior version of TestCafe we were using wasn't very good at showing the diff between two objects when we tried to compare them using `.eql`. It would look like this:

image

which wasn't very useful because it didn't show us which property was different. This is why I had to output both serialized objects like this when the test failed so that the developer could at least compare the objects visually. The latest version added an enhancement that shows a better diff of the object:

image

So now we no longer need to log the stringified objects. 🎉

**Pull request merged**
